### DISCUSSION TITLE
`Reflection` not support custom xaml control

### DISCUSSION THREAD

**Discussion by SongOfYouth** [unoplatform/uno#12858] [Q&A]
### Current behavior
I have a custom class named `BasePage` (empty class) which inherited the `Page` class, and another class named `MainPage` inherited the `BasePage`, when i create a instance of `MainPage` by reflection,i got a error: Microsoft.UI.Xaml.Markup.XamlParseException:“XAML parsing failed.”, but if i refer the project which the `MainPage`  in,and reflect it or create it by `new`,it works.

### Works on UWP/WinUI

None

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### Affected platforms

Windows

### IDE

Visual Studio 2022

**Comment by SongOfYouth** [comment]
trailed some times, I find the reason. The compiler will generate a `XamlTypeInfo.g.cs` file when building, which the `XamlTypeInfoProvider` of other project will be builded in when reference other project with `ProjectReference`, and this section is missing when reflect a assembly directly. My solution code is :
in `App.cs`:
```csharp
var appProviderProperty = this.GetType().GetProperty("_AppProvider", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
var appProvider = appProviderProperty!.GetValue(this)!;
var appProviderProviderProperty = appProvider.GetType().GetProperty("Provider", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
var appProviderProvider = appProviderProviderProperty!.GetValue(appProvider);
var othersProperty = appProviderProvider!.GetType().GetProperty("OtherProviders", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
var otherProviders = othersProperty!.GetValue(appProviderProvider) as List<Microsoft.UI.Xaml.Markup.IXamlMetadataProvider>;
var originType = Assembly.Load("MyAssembly").GetType("MyAssembly.MyAssembly_XamlTypeInfo.XamlMetaDataProvider");
var provider = Activator.CreateInstance(originType!) as Microsoft.UI.Xaml.Markup.IXamlMetadataProvider;
otherProviders!.Add(provider!);
```
Is it possible to provide a frendly way to do it in future?

**Reply by jeromelaban** [reply]
I'm not sure to understand what you're trying to do, and the error you had originally. If you can provide more details about the scenario you're trying to get working, it will help.

 Note that if the issue happens when running the Windows head, this will be a [WinUI issue](https://github.com/microsoft/microsoft-ui-xaml), not an Uno issue.

**Reply by SongOfYouth** [reply]
I think it's possibly a WinUI issue. as the [reply](https://github.com/unoplatform/uno/discussions/12855#discussioncomment-6421707) saying, I want to load a control(eg. Page inherited from a custom page class) from assembly file at runtime, When I reflect it directly and create a instance, I get the error above, and when i reference the project of this assembly by `ProjectReference` and reflect it, it works! I have found that, in the  `XamlTypeInfo.g.cs` file, the `XamlTypeInfoProvider` of the assembly I reflected will be add to the **OtherProviders** collection when reference it by `ProjectReference` , I think that's the difference.
![image](https://github.com/unoplatform/uno/assets/30856466/edb51505-c8cd-4e47-8076-f57c1cfd540b)
